plugins/syntax/number.ml

```diff
@@ -0,0 +1,479 @@
+   [to_post] translation (c.f., interp/notation.mli) for the numeral
+   notation to parse/print type [ty_ind]. This translation is the
+   identity ([ToPostCopy]) except that it checks ([ToPostCheck]) that
+   the parameters of the inductive type [ty_ind] match the ones given
+   in [params]. *)
+let elaborate_to_post_params env sigma ty_ind params =
+  let to_post_for_constructor indc =
+    let sigma, c = match indc with
+      | GlobRef.ConstructRef c ->
+         let sigma,c = Evd.fresh_constructor_instance env sigma c in
+         sigma, Constr.mkConstructU c
+      | _ -> assert false in  (* c.f. get_constructors *)
+    let args, t = get_type env sigma c in
+    let params_indc = match Constr.kind t with
+      | Constr.App (_, a) -> Array.to_list a | _ -> [] in
+    let sz = List.length args in
+    let a = Array.make sz ToPostCopy in
+    if List.length params <> List.length params_indc then error_params ty_ind;
+    List.iter2 (fun param param_indc ->
+        match param, Constr.kind param_indc with
+        | Some p, Constr.Rel i when i <= sz -> a.(sz - i) <- ToPostCheck p
+        | _ -> ())
+      params params_indc;
+    indc, indc, Array.to_list a in
+  let pt_refs = get_constructors ty_ind in
+  let to_post_0 = List.map to_post_for_constructor pt_refs in
+  let to_post =
+    let only_copy (_, _, args) = List.for_all ((=) ToPostCopy) args in
+    if (List.for_all only_copy to_post_0) then [||] else [|to_post_0|] in
+  to_post, pt_refs
+
+(* [elaborate_to_post env sigma ty_name ty_ind l] builds the [to_post]
+   translation (c.f., interp/notation.mli) for the number notation to
+   parse/print type [ty_name] through the inductive [ty_ind] according
+   to the pairs [constant, constructor] in the list [l]. *)
+let elaborate_to_post_via env sigma ty_name ty_ind l =
+  let sigma, ty_name =
+    locate_global_sort_inductive_or_constant sigma ty_name in
+  let ty_ind = Constr.mkInd ty_ind in
+  (* Retrieve constants and constructors mappings and their type.
+     For each constant [cnst] and inductive constructor [indc] in [l], retrieve:
+     * its location: [lcnst] and [lindc]
+     * its GlobRef: [cnst] and [indc]
+     * its type: [tcnst] and [tindc] (decomposed in product by [get_type] above)
+     * [impls] are the implicit arguments of [cnst] *)
+  let l =
+    let read (consider_implicits, cnst, indc) =
+      let lcnst, lindc = cnst.CAst.loc, indc.CAst.loc in
+      let cnst, ccnst = locate_global_constructor_inductive_or_constant cnst in
+      let indc, cindc =
+        let indc = Smartlocate.global_constructor_with_alias indc in
+        GlobRef.ConstructRef indc, Constr.mkConstruct indc in
+      let get_type c =
+        (* ignore parameters of inductive types *)
+        let rm_params c = match Constr.kind c with
+          | Constr.App (c, _) when Constr.isInd c -> c
+          | _ -> c in
+        let l, t = get_type env sigma c in
+        List.map (fun (n, c) -> n, rm_params c) l, rm_params t in
```

Comment:
Why do you rm_params on the second components of l?  Shouldn't they all be constructors, and hence rm_params would be a no-op?

Reply:
Oh, you rebind l just above this.  I think you should avoid shadowing except in cases where using the previous value is a genuine error, like for env and sigma.  And it's a bit confusing that this let-bound get_type function calls another function called get_type which is not itself....

Reply:
fixed